File: 0298-Binary Tree Longest Consecutive Sequence/0298-Binary Tree Longest Consecutive Sequence.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 class Solution {
public:
    int longestConsecutive(TreeNode* root) {
        int maxLen = 0;
        helper(root, maxLen);
        return maxLen;
    }
private:
    int helper(TreeNode* root, int& maxLen) {
        if (root == nullptr) {
            return 0;
        }
        
        int leftLen = helper(root->left, maxLen);
        int rightLen = helper(root->right, maxLen);
        int len = 1;
        if (root->left && root->val == root->left->val - 1) {
            len = 1 + leftLen;
        }
        
        if (root->right && root->val == root->right->val - 1) {
            len = max(len, 1 + rightLen);
        }
        
        if (len > maxLen) {
            maxLen = len;
        }
        
        return len;
    }
};

File: 0201-0300/0298-Binary Tree Longest Consecutive Sequence/0298-Binary Tree Longest Consecutive Sequence.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def longestConsecutive(self, root: TreeNode) -> int:
        maxLen = 0
        def dfs(root):
            if root is None:
                return 0
            
            currLen = 1
            leftLen = dfs(root.left)
            rightLen = dfs(root.right)
            if root.left and root.left.val == root.val + 1:
                currLen += leftLen
            if root.right and root.right.val == root.val + 1:
                currLen = max(currLen, 1 + rightLen)
            nonlocal maxLen
            maxLen = max(maxLen, currLen)
            return currLen

        dfs(root)
        return maxLen